### PR TITLE
Refactor BlockCompare block to use React hooks

### PR DESCRIPTION
**Pull request by truchot**
## Description
Related #22890

## How has this been tested?
1. Edit page or post
2. Go to editor code
3. Code wrong block
4. Save content
5. Refresh page
6. Message "This block contains unexpected or invalid content." will appear
7. Click on "Resolve" button

## Types of changes
Refactor BlockCompare to use React hooks.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-compare/index.js [hunk 1]
```diff
@@ -11,16 +11,21 @@ import { diffChars } from 'diff/lib/diff/character';
...
-class BlockCompare extends Component {
-	getDifference( originalContent, newContent ) {
+function BlockCompare( {
+	block,
+	onKeep,
+	onConvert,
+	convertor,
+	convertButtonText,
+} ) {
+	const getDifference = ( originalContent, newContent ) => {
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-compare/index.js:28]: Ideally these are named functions so it's easier to debug

File: packages/block-editor/src/components/block-compare/index.js [hunk 2]
```diff
@@ -60,45 +65,36 @@ class BlockCompare extends Component {
...
-			convertor,
-			convertButtonText,
-		} = this.props;
-		const original = this.getOriginalContent( block );
-		const converted = this.getConvertedContent( convertor( block ) );
-		const difference = this.getDifference(
-			original.rawContent,
-			converted.rawContent
-		);
+	const original = getOriginalContent();
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-compare/index.js:70]: Why does this need to be a function?

Comment (truchot) [reply packages/block-editor/src/components/block-compare/index.js:70]: Yes I can remove this function it's no longer relevant.

File: packages/block-editor/src/components/block-compare/index.js (hunk 1 above)
Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-compare/index.js:28]: In other words, this should be a standard function declaration, e.g. `function getDifference(...`.

Comment (truchot) [reply packages/block-editor/src/components/block-compare/index.js]: Like this ? https://github.com/WordPress/gutenberg/pull/22909/commits/26490e9903ce830315df84a15337fbb24ec15cd1

Comment (ZebulanStanphill) [reply packages/block-editor/src/components/block-compare/index.js]: Yep! :+1:

File: packages/block-editor/src/components/block-compare/index.js [hunk 3]
```diff
@@ -11,16 +11,21 @@ import { diffChars } from 'diff/lib/diff/character';
...
 import { getSaveContent, getSaveElement } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import BlockView from './block-view';
 
-class BlockCompare extends Component {
-	getDifference( originalContent, newContent ) {
+function BlockCompare( {
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/block-compare/index.js:21]: Nitpick: you could do `export default function` here:
```suggestion
export default function BlockCompare( {
```

**Review by ZebulanStanphill — APPROVED** [review]
LGTM (looks good to merge)!

**Pull request merged**